torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,731 @@
+
+    NCCL_CHECK(ncclReduce(data[i].data_ptr(),
+                          data[i].data_ptr(),
+                          data[i].numel(),
+                          _getNcclDataType(data[i].type().scalarType()),
+                          ncclOp[operation],
+                          dstRank * data.size(),
+                          (*comms)[i],
+                          stream));
+    THCudaCheck(cudaEventRecord((*events)[i], stream));
+  }
+  NCCL_CHECK(ncclGroupEnd());
+
+  cudaFreeMutexLock.unlock();
+}
+
+void DataChannelNccl::reduce(at::Tensor& data,
+                             THDReduceOp operation,
+                             rank_type dstRank,
+                             THDGroup groupId) {
+
+  std::vector<at::Tensor> dataVec;
+  dataVec.push_back(data);
+  reduce(dataVec, operation, dstRank, groupId);
+}
+
+void DataChannelNccl::broadcast(std::vector<at::Tensor>& data,
+                                rank_type srcRank,
+                                THDGroup groupId) {
+
+  std::unique_lock<std::mutex> channelLock(_mutex);
+
+  // Check the tensor vector for consistency
+  if (!_tensorCheckHelper(data, data)) {
+    return;
+  }
+  _checkGroupIdValid(groupId);
+
+  auto commsAndEvents = _getNcclCommsAndEvents(data, groupId);
+  auto comms = commsAndEvents.first;
+  auto events = commsAndEvents.second;
+
+  // Guard GPU device
+  AutoGPU gpuGuard;
+
+  std::unique_lock<std::mutex> cudaFreeMutexLock(
+      *(THCCachingAllocator_getCudaFreeMutex()));
+
+  NCCL_CHECK(ncclGroupStart());
+  for (size_t i = 0; i < data.size(); ++i) {
+
+    gpuGuard.setDevice(data[i].get_device());
+    auto stream = THCState_getCurrentStream(THDGetCudaState());
+
+    NCCL_CHECK(ncclBcast(data[i].data_ptr(),
+                         data[i].numel(),
+                         _getNcclDataType(data[i].type().scalarType()),
+                         srcRank * data.size(),
+                         (*comms)[i],
+                         stream));
+    THCudaCheck(cudaEventRecord((*events)[i], stream));
+  }
+  NCCL_CHECK(ncclGroupEnd());
+
+  cudaFreeMutexLock.unlock();
+}
+
+
+void DataChannelNccl::broadcast(at::Tensor& data,
+                                rank_type srcRank,
+                                THDGroup groupId) {
+
+  std::vector<at::Tensor> dataVec;
+  dataVec.push_back(data);
+  broadcast(dataVec, srcRank, groupId);
+}
+
+
+void DataChannelNccl::barrier(THDGroup groupId) {
+
+  std::unique_lock<std::mutex> channelLock(_mutex);
+
+  _checkGroupIdValid(groupId);
+
+  if (_ncclCommsAndEvents.find(groupId) == _ncclCommsAndEvents.end() ||
+      _groupDevices.find(groupId) == _groupDevices.end()) {
+    return;
+  }
+
+  auto devices = getDevicesList(_groupDevices[groupId]);
+
+  // Guard GPU device
+  AutoGPU gpuGuard;
+
+  int idx = 0;
+  // Synchronize on the CUDA events
+  for (auto& event : *(_ncclCommsAndEvents[groupId].second)) {
+    gpuGuard.setDevice(devices[idx++]);
+    THCudaCheck(cudaEventSynchronize(event));
+  }
+}
```

Comment:
That doesn't really implement a distributed barrier. It only syncs with locally queued kernels

Reply:
It will for NCCL